core/src/main/java/com/datastax/oss/driver/internal/core/metadata/NodeStateEvent.java

```diff
@@ -53,14 +55,19 @@ public static NodeStateEvent removed(DefaultNode node) {
    */
   public final NodeState newState;
 
-  public final DefaultNode node;
+  private final WeakReference<DefaultNode> node;
```

Comment:
@absurdfarce I will try to revert the changes and see whether it still leaks, but:
There are two `WeakHashMap` held by `ControlConnection`:
```java
    private final Map<Node, DistanceEvent> lastDistanceEvents = new WeakHashMap<>();
    private final Map<Node, NodeStateEvent> lastStateEvents = new WeakHashMap<>();
```
However, both DistanceEvent and NodeStateEvent hold strong references to the associated node 
From the [doc](https://docs.oracle.com/javase/8/docs/api/java/util/WeakHashMap.html) of WeakHashMap:

> Implementation note: The value objects in a WeakHashMap are held by ordinary strong references. Thus care should be taken to ensure that value objects do not strongly refer to their own keys, either directly or indirectly, since that will prevent the keys from being discarded.

In this case `DistanceEvent` and `NodeStateEvent` are the values that hold strong references to the keys. Therefore, I think it may leak references.

Reply:
Tested out. If `NodeStateEvent` and `DistanceEvent` hold strong references to `node`, then objects of these two kinds, together with `DefaultNode`, will not be garbage collected.

Reply:
Confirmed that if we put `lastStateEvent` and `lastDistanceEvent` as two fields of `DefaultNode`, then it will work. `DefaultNode` <-> `NodeStateEvent` only point to each other, same for `DistanceEvent`. Objects of all three classes will be collected. However, I understand this is not an ideal solution.

Reply:
What about this:

```java
    private final Map<Node, NodeDistance> lastDistanceEvents = new WeakHashMap<>();
    private final Map<Node, NodeState> lastStateEvents = new WeakHashMap<>();
```